### PR TITLE
UCT/IB: Move res_domain to accelerated TLs

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9065/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 1]
```diff
@@ -301,6 +301,38 @@ struct uct_ib_mlx5_atomic_masked_fadd64_seg {
...
+    struct ibv_pd               *pd;
+    struct ibv_pd               *real_pd;
+#endif
+} uct_ib_mlx5_iface_res_domain_t;
+
+
+/**
+ *  MLX5 common iface part
+ */
+typedef struct uct_ib_mlx5_iface {
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: uct_ib_mlx5_iface_common

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 2]
```diff
@@ -234,6 +234,7 @@ typedef union uct_rc_mlx5_dm_copy_data {
 
 typedef struct uct_rc_mlx5_iface_common {
     uct_rc_iface_t                   super;
+    uct_ib_mlx5_iface_t              mlx5;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.h]: mlx5_common

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -923,27 +804,16 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
...
-                                                  init_attr->res_domain_key,
-                                                  uct_ib_iface_res_domain_t,
-                                                  uct_ib_iface_res_domain_cmp,
-                                                  uct_ib_iface_res_domain_init,
-                                                  self);
-        if (UCS_PTR_IS_ERR(self->res_domain)) {
-            status = UCS_PTR_STATUS(self->res_domain);
-            goto err_free_path_bits;
-        }
+    status = self->ops->setup_iface(self);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: rename to init_res_domain / cleanaup_res_domain

File: src/uct/ib/base/ib_iface.c [hunk 4]
```diff
@@ -552,7 +545,7 @@ ucs_status_t uct_ib_iface_create_qp(uct_ib_iface_t *iface,
 #elif HAVE_DECL_IBV_CREATE_QP_EX
     qp = ibv_create_qp_ex(dev->ibv_context, &attr->ibv);
 #else
-    qp = ibv_create_qp(uct_ib_iface_qp_pd(iface), &attr->ibv);
+    qp = ibv_create_qp(iface->ops->get_qp_pd(iface), &attr->ibv);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: use iface's device pd

File: src/uct/ib/base/ib_iface.c [hunk 5]
```diff
@@ -502,17 +502,10 @@ void uct_ib_iface_fill_attr(uct_ib_iface_t *iface, uct_ib_qp_attr_t *attr)
 
 #if HAVE_DECL_IBV_EXP_CREATE_QP
     attr->ibv.comp_mask           = IBV_EXP_QP_INIT_ATTR_PD;
-    attr->ibv.pd                  = uct_ib_iface_qp_pd(iface);
+    attr->ibv.pd                  = iface->ops->get_qp_pd(iface);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: if PD flag not set in comp_mask, then set default PD
and set upstream td/exp resd in mlx5_fill_attr

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 6]
```diff
@@ -140,6 +140,143 @@ ucs_status_t uct_ib_mlx5_get_cq(struct ibv_cq *cq, uct_ib_mlx5_cq_t *mlx5_cq)
...
 }
 
+static int
+uct_ib_mlx5_iface_res_domain_cmp(uct_ib_mlx5_iface_res_domain_t *res_domain,
+                                 uct_ib_md_t *md, uct_priv_worker_t *worker)
+{
+#if HAVE_IBV_EXP_RES_DOMAIN
+    return res_domain->ibv_domain->context == md->dev.ibv_context;
+#elif HAVE_DECL_IBV_ALLOC_TD
+    return res_domain->real_pd == md->pd;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c]: compare with md->context == md->pd instead of keeping real_pd

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9185/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/base/ib_iface.h [hunk 7]
```diff
@@ -147,21 +147,11 @@ struct uct_ib_iface_ops {
                                              ucs_status_t status);
     ucs_status_t            (*create_qp)(uct_ib_iface_t *iface, uct_ib_qp_attr_t *attr,
                                          struct ibv_qp **qp_p);
+    ucs_status_t            (*setup_iface)(uct_ib_iface_t *iface);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: rename to init_res_domain / cleanup_res_domain

File: src/uct/ib/base/ib_iface.c [hunk 8]
```diff
@@ -1308,3 +1178,7 @@ ucs_status_t uct_ib_iface_arm_cq(uct_ib_iface_t *iface,
     return UCS_OK;
 }
 
+struct ibv_pd *uct_ib_iface_qp_pd(uct_ib_iface_t *iface)
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: remove?

File: src/uct/ib/base/ib_iface.h [hunk 9]
```diff
@@ -533,19 +521,7 @@ void uct_ib_iface_fill_ah_attr_from_addr(uct_ib_iface_t *iface,
...
-
-    pd = uct_ib_iface_md(iface)->pd;
-#if HAVE_DECL_IBV_ALLOC_TD
-    if (iface->res_domain && iface->res_domain->ibv_domain) {
-        pd = iface->res_domain->ibv_domain;
-    }
-#endif
-    return pd;
-}
+struct ibv_pd *uct_ib_iface_qp_pd(uct_ib_iface_t *iface);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: remove?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9207/ for details (Mellanox internal link).

**Pull request merged**